scripts/zephyr_module.py

```diff
@@ -395,33 +398,43 @@ def parse_modules(zephyr_base, modules=None, extra_modules=None):
     if extra_modules is None:
         extra_modules = []
 
-    Module = namedtuple('Module', ['project', 'meta', 'depends'])
-    # dep_modules is a list of all modules that has an unresolved dependency
-    dep_modules = []
-    # start_modules is a list modules with no depends left (no incoming edge)
-    start_modules = []
-    # sorted_modules is a topological sorted list of the modules
-    sorted_modules = []
+    # Process the reversed module list and ignore modules with duplicated names
+    # to allow for overriding the built-in modules by the extra modules.
+
+    all_modules = []
+    all_module_names = set()
 
-    for project in modules + extra_modules:
+    for project in reversed(modules + extra_modules):
         # Avoid including Zephyr base project as module.
         if project == zephyr_base:
             continue
 
-        meta = process_module(project)
-        if meta:
-            section = meta.get('build', dict())
-            deps = section.get('depends', [])
-            if not deps:
-                start_modules.append(Module(project, meta, []))
-            else:
-                dep_modules.append(Module(project, meta, deps))
-        elif project in extra_modules:
+        module = process_module(project)
+        if module and module.meta['name'] not in all_module_names:
+            all_modules.append(module)
+            all_module_names.add(module.meta['name'])
```

Comment:
you don't need to over complicate this.

Just create an empty dictionary and use `meta` name as key.
Then you can update it based on the `meta` name, and last update will win.

Like this:
```diff
--- a/scripts/zephyr_module.py
+++ b/scripts/zephyr_module.py
@@ -396,6 +396,7 @@ def parse_modules(zephyr_base, modules=None, extra_modules=None):
         extra_modules = []
 
     Module = namedtuple('Module', ['project', 'meta', 'depends'])
+    modules_dict = {}
     # dep_modules is a list of all modules that has an unresolved dependency
     dep_modules = []
     # start_modules is a list modules with no depends left (no incoming edge)
@@ -410,16 +411,19 @@ def parse_modules(zephyr_base, modules=None, extra_modules=None):
 
         meta = process_module(project)
         if meta:
-            section = meta.get('build', dict())
-            deps = section.get('depends', [])
-            if not deps:
-                start_modules.append(Module(project, meta, []))
-            else:
-                dep_modules.append(Module(project, meta, deps))
+            modules_dict.update({meta.get('name'): {'project': project, 'meta': meta}})
         elif project in extra_modules:
             sys.exit(f'{project}, given in ZEPHYR_EXTRA_MODULES, '
                      'is not a valid zephyr module')
 
+    for name, value in modules_dict.items():
+        section = value['meta'].get('build', dict())
+        deps = section.get('depends', [])
+        if not deps:
+            start_modules.append(Module(value['project'], value['meta'], []))
+        else:
+            dep_modules.append(Module(value['project'], value['meta'], deps))
+
     # This will do a topological sort to ensure the modules are ordered
     # according to dependency settings.
     while start_modules:
```

Reply:
I wanted to preserve the order or items in the module list. I was concerned that if I change it, some projects will start failing in case some modules don't have dependencies properly configured. Should I not worry? :)

Reply:
> I was concerned that if I change it, some projects will start failing in case some modules don't have dependencies properly configured. Should I not worry? :)

Correct, you should not worry.
We don't guarantee a specific order although in practice we use the order given by the argument / or provided by `west` when `west` is used.

With that said, the beauty of `dict.update()` is that it doesn't change the order, it updates at the index given.

For example with vanilla Zephyr you have this order:
```
...
10 "infineon":"/projects/github/ncs/modules/hal/infineon":"/projects/github/ncs/modules/hal/infineon"
11 "microchip":"/projects/github/ncs/modules/hal/microchip":"/projects/github/ncs/modules/hal/microchip"
12 "hal_nordic":"/projects/github/ncs/modules/hal/nordic":"${ZEPHYR_HAL_NORDIC_CMAKE_DIR}"
13 "nuvoton":"/projects/github/ncs/modules/hal/nuvoton":"/projects/github/ncs/modules/hal/nuvoton"
14 "hal_nxp":"/projects/github/ncs/modules/hal/nxp":"${ZEPHYR_HAL_NXP_CMAKE_DIR}"
...
```
and if I want to use my custom hal Nordic as: `-DZEPHYR_EXTRA_MODULES=(pwd)/../modules/hal/custom_hal_nordic/
` I'll get this:
```
...
 10 "infineon":"/projects/github/ncs/modules/hal/infineon":"/projects/github/ncs/modules/hal/infineon"
 11 "microchip":"/projects/github/ncs/modules/hal/microchip":"/projects/github/ncs/modules/hal/microchip"
 12 "hal_nordic":"/projects/github/ncs/zephyr/../modules/hal/custom_hal_nordic":"/projects/github/ncs/modules/hal/custom_hal_nordic"
 13 "nuvoton":"/projects/github/ncs/modules/hal/nuvoton":"/projects/github/ncs/modules/hal/nuvoton"
 14 "hal_nxp":"/projects/github/ncs/modules/hal/nxp":"${ZEPHYR_HAL_NXP_CMAKE_DIR}"
...
```
so as you see, order is preserved when injecting my custom version :wink:

Reply:
ok, thanks. Changed it.